reflect/src/main/java/dagger/reflect/Linker.java

```diff
@@ -27,10 +27,7 @@ private Linker(Scope scope) {
   }
 
   @Nullable LinkedBinding<?> find(Key key) {
-    Binding binding = scope.bindings.get(key);
-    if (binding instanceof UnlinkedBinding) {
-      return performLinking(key, (UnlinkedBinding) binding);
```

Comment:
So this is why the cycle test starts failing. The linker needs to be responsible for performing all linking. Right now you probably infinitely recurse on that test?